include/verilated.cpp

```diff
@@ -637,26 +637,85 @@ std::string VL_DECIMAL_NW(int width, WDataInP lwp) VL_MT_SAFE {
     return output;
 }
 
-std::string _vl_vsformat_time(char* tmp, double ld, bool left, size_t width) {
-    // Double may lose precision, but sc_time_stamp has similar limit
-    std::string suffix = Verilated::threadContextp()->impp()->timeFormatSuffix();
-    int userUnits = Verilated::threadContextp()->impp()->timeFormatUnits();  // 0..-15
-    int fracDigits = Verilated::threadContextp()->impp()->timeFormatPrecision();  // 0..N
-    int prec = Verilated::threadContextp()->timeprecision();  // 0..-15
-    int shift = prec - userUnits + fracDigits;  // 0..-15
-    double shiftd = vl_time_multiplier(shift);
-    double scaled = ld * shiftd;
-    const double fracDiv = vl_time_multiplier(fracDigits);
-    const double whole = scaled / fracDiv;
+template <typename T>
+std::string _vl_vsformat_time(char* tmp, T ld, int timeunit, bool left, size_t width) {
+    const std::string suffix = Verilated::threadContextp()->impp()->timeFormatSuffix();
+    const int userUnits = Verilated::threadContextp()->impp()->timeFormatUnits();  // 0..-15
+    const int fracDigits = Verilated::threadContextp()->impp()->timeFormatPrecision();  // 0..N
+    const int shift = -userUnits + fracDigits + timeunit;  // 0..-15
     int digits = 0;
-    if (!fracDigits) {
-        digits = VL_SNPRINTF(tmp, VL_VALUE_STRING_MAX_WIDTH, "%.0f%s", whole, suffix.c_str());
+    if (std::numeric_limits<T>::is_integer) {
+        static const vluint64_t pow10[20] = {
+            1ULL,
+            10ULL,
+            100ULL,
+            1000ULL,
+            10000ULL,
+            100000ULL,
+            1000000ULL,
+            10000000ULL,
+            100000000ULL,
+            1000000000ULL,
+            10000000000ULL,
+            100000000000ULL,
+            1000000000000ULL,
+            10000000000000ULL,
+            100000000000000ULL,
+            1000000000000000ULL,
+            10000000000000000ULL,
+            100000000000000000ULL,
+            1000000000000000000ULL,
+        };
```

Comment:
Perhaps cleaner if factored into a little vl_time_pow10 function, put it next to vl_time_multiplier please.